Fix signing for URL query parameters with '='

sblom opened the pull request:
When a URL query parameter contains `=` (%3D), `deParam()` drops everything after the second `=`.

This change preserves all of the contents after the first `=` (including additional `=`).

sblom commented:
I signed it!

On Tue, Oct 20, 2015 at 2:34 PM, googlebot <notifications@github.com> wrote:

> Thanks for your pull request. It looks like this may be your first
> contribution to a Google open source project, in which case you'll need to
> sign a Contributor License Agreement (CLA).
>
> [image: :memo:] *Please visit https://cla.developers.google.com/
> <https://cla.developers.google.com/> to sign.*
>
> Once you've signed, please reply here (e.g. I signed it!) and we'll
> verify. Thanks.
> ------------------------------
>
>    - If you've already signed a CLA, it's possible we don't have your
>    GitHub username or you're using a different email address. Check your
>    existing CLA data <https://cla.developers.google.com/clas> and verify
>    that your email is set on your git commits
>    <https://help.github.com/articles/setting-your-email-in-git/>.
>    - If you signed the CLA as a corporation, please let us know the
>    company's name.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/googlesamples/apps-script-oauth1/pull/15#issuecomment-149710036>
> .
>

erickoledadevrel commented:
I'm fairly certain this isn't the correct solution to your problem. My guess is that you are using escaping:false and calling the Twitter API? If you look at the Twitter sample, you'll see that you don't need to do your own escaping anymore.

sblom commented:
Nah. This isn't related to doing my own escaping--it's for a GET, not a POST.

I'm trying to run a query against the QuickBooks API:
> https://sandbox-quickbooks.api.intuit.com/v3/company/1420297955/query?query=select%20%2A%20from%20Account%20where%20Classification%3D%27Expense%27

(Unescaped, that's `select * from Account where Classification='Expense'`.)

`deParam()` is returning

    {
        query: 'select * from Account where Classification'
    }

as the normalized query parameters to be signed. (Notice that `='Expense'` was dropped by the `.split('=')` as it currently stands.)

This results in an invalid OAuth signature. The fact that the `.split('=')` is splitting on %3D in addition to an unescaped '=' suggests to me that my workaround couldn't possible be to do _less_ escaping...

erickoledadevrel commented:
Ah, thanks for the clarification. It looks like the problem was recently fixed in the library that Signer.gs is based on:

https://github.com/ddo/oauth-1.0a/commit/4768db7b2380271165735f7a66b5cbaffc53ead0

I think the solution there is preferred, where decodeURIComponent is called after the string has been split, not before. Can you update your commit to use the latest version of the deParam() function?

sblom commented:
Okay--incorporated upstream's latest implementation. I agree that it's cleaner.

michaes commented:
I still see this error happening with QuickBooks (I am using version 13 of this library)

For the URL https://quickbooks.api.intuit.com/v3/company/123145856926847/reports/ProfitAndLoss?%20where%20accounting_period%3D'Cash'

I get the below error 
Status : 401.0, 
Response : "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>\n<IntuitResponse time=\"2016-10-20T10:26:52.545-07:00\" xmlns=\"http://schema.intuit.com/finance/v3\">\n    <Fault type=\"AUTHENTICATION\">\n        <Error code=\"3200\">\n            <Message>message=ApplicationAuthenticationFailed; errorCode=003200; statusCode=401</Message>\n        </Error>\n    </Fault>\n</IntuitResponse>\n"


I am blocked with this, appreciate if you could help me to resolve this issue.
thanks

erickoledadevrel commented:
Good catch, it looks like I mistakenly reverted the fix in a following commit. I've just pushed out version 14 of the library that fixes the issue again.
